File: src/main/java/com/flaghacker/uttt/bots/MMBot.java
package com.flaghacker.uttt.bots;

import com.flaghacker.uttt.common.Board;
import com.flaghacker.uttt.common.Bot;
import com.flaghacker.uttt.common.Coord;
import com.flaghacker.uttt.common.Player;
import com.flaghacker.uttt.common.Timer;

public class MMBot implements Bot
{
	private final int depth;

	public MMBot(int depth)
	{
		this.depth = depth;
	}

	@Override
	public Coord move(Board board, Timer timer)
	{
		if (board.nextPlayer() == Player.ENEMY)
			board = board.flip();

		double bestValue = 0;
		Coord bestMove = null;

		for (Coord move : board.availableMoves())
		{
			Board next = board.copy();
			next.play(move);

			double negaMax = -negaMax(next, depth, -1);

			if (bestMove == null || negaMax > bestValue)
			{
				bestValue = negaMax;
				bestMove = move;
			}
		}

		return bestMove;
	}

	private double negaMax(Board board, int depth, int player)
	{
		if (depth == 0 || board.isDone())
			return player * value(board);

		double bestValue = Double.NEGATIVE_INFINITY;

		for (Coord move : board.availableMoves())
		{
			Board next = board.copy();
			next.play(move);

			double value = -negaMax(next, depth - 1, -player);
			bestValue = Math.max(bestValue, value);
		}

		return bestValue;
	}

	private double value(Board board)
	{
		if (board.isDone())
			return Double.POSITIVE_INFINITY * playerSign(board.wonBy());

		double value = 0;

		for (Coord coord : Coord.list())
			value += TILE_VALUE * tileFactor(coord.os()) * tileFactor(coord.om()) * playerSign(board.tile(coord));

		for (int om = 0; om < 9; om++)
			value += MACRO_VALUE * tileFactor(om) * playerSign(board.macro(om));

		return value;
	}

	private static final double TILE_VALUE = 1;
	private static final double MACRO_VALUE = 10e9;

	private static final double CENTER_FACTOR = 4;
	private static final double CORNER_FACTOR = 3;
	private static final double EDGE_FACTOR = 1;

	private int playerSign(Player player)
	{
		switch (player)
		{
			case NEUTRAL:
				return 0;
			case PLAYER:
				return 1;
			case ENEMY:
				return -1;
		}

		throw new IllegalStateException();
	}

	private double tileFactor(int o)
	{
		int x = o % 3;
		int y = o / 3;

		if (x == 1 && y == 1)
			return CENTER_FACTOR;
		if (x == 1 || y == 1)
			return EDGE_FACTOR;
		return CORNER_FACTOR;
	}

	@Override
	public String toString()
	{
		return "MMBot";
	}
}

File: src/main/kotlin/bots/MMBot.kt
package bots

import common.Board
import common.Bot
import common.Coord
import kotlin.Float.Companion.NEGATIVE_INFINITY
import kotlin.math.max

private const val TILE_FACTOR = 1
private const val MACRO_FACTOR = 32

private const val CENTER_FACTOR = 4
private const val CORNER_FACTOR = 2
private const val EDGE_FACTOR = 1

class MMBot(private val depth: Int) : Bot {
	init {
		require(depth >= 1) { "depth must be >= 1, was $depth" }
	}

	override fun move(board: Board): Coord {
		if (board.isDone) throw Exception("No moves remain")

		return negaMax(
			board, 0.0F, depth,
			java.lang.Float.NEGATIVE_INFINITY, java.lang.Float.POSITIVE_INFINITY,
			if (board.nextPlayX) 1 else -1
		).move
	}

	private class ValuedMove(val move: Coord, val value: Float)

	private fun negaMax(board: Board, cValue: Float, depth: Int, a: Float, b: Float, player: Int): ValuedMove {
		if (depth == 0 || board.isDone) return ValuedMove(board.lastMove, player * cValue)

		var bestValue = NEGATIVE_INFINITY
		var bestMove: Coord? = null
		var newA = a

		var stopSearch = false
		board.forAvailableMoves { move ->
			if (!stopSearch){
				val child = board.copy()

				//Calculate the new score
				var childValue = cValue + TILE_FACTOR * factor(move.toInt() and 0b1111) * factor((move.toInt() shr 4) and 0b1111) * player
				if (child.play(move)) {
					if (child.isDone) childValue = java.lang.Float.POSITIVE_INFINITY * player
					else childValue += MACRO_FACTOR * factor((move.toInt() shr 4) and 0b1111) * player
				}

				//Check if the (global) value of this child is better than the previous best child
				val value = -negaMax(child, childValue, depth - 1, -b, -newA, -player).value
				if (value > bestValue || bestMove == null) {
					bestValue = value
					bestMove = child.lastMove
				}
				newA = max(newA, value)
				if (newA >= b) stopSearch = true
			}
		}

		return ValuedMove(bestMove!!, bestValue)
	}

	private fun factor(os: Int) = when {
		os == 4 -> CENTER_FACTOR
		os % 2 == 0 -> CORNER_FACTOR
		else -> EDGE_FACTOR
	}

	override fun toString() = "MMBot(d=$depth)"
}